### PR TITLE
Adds lint rules to MSVC

### PR DESCRIPTION
**Pull request by cugone**
MSVC version of #246

Adds Code Analysis rules (lint rules) for MSVC by adding checks for all the Core Guidelines. The custom `.ruleset` file is required to select multiple rule sets to follow.

There are also `Microsoft Native Minimum Rules` and `Microsoft Native Recommended Rules`, the `Minimum` is recommended for all custom rulesets and `Recommended` is for use with Visual Studio Professional Edition. They are Microsoft-specific so I decided to defer using them.

### CODE REVIEW

**Comment by cugone** [comment]
@DanRStevens 

The default CA settings and rules are `Microsoft Minimum Rules` for Debug builds and `Disabled` for non-Debug builds. I've replaced them with All Core Guidelines for all builds.

Should I turn on `Minimum` for all builds as well as the Core Guidelines?

**Comment by DanRStevens** [comment]
Woah, the builds just took 13-14 minutes per configuration, and produced over 7000 lines of output. That's overwhelming. AppVeyor only shows the most recent 2000 lines, so we're missing most of the output.

I would like to use increased linting support, but I think we're going to have to roll this one out in stages. It's not practical to merge as is.

Maybe we can experiment with enabling just one or two sets of warning in a branch, and focus on fixing those.

----

I'm uncertain about replacing `Microsoft Native Minimum Rules`. These might be a different set than the core guidelines. Maybe we'd like to enable both sets of warnings? (Assuming they don't conflict with each other, each requiring something the other will warn about).

File: proj/vs2019/ophd.vcxproj [hunk 1]
```diff
@@ -89,7 +93,7 @@
       <PrecompiledHeader />
       <WarningLevel>Level4</WarningLevel>
       <DebugInformationFormat>ProgramDatabase</DebugInformationFormat>
-      <EnablePREfast>false</EnablePREfast>
+      <EnablePREfast>true</EnablePREfast>
```
Comment (DanRStevens) [inline proj/vs2019/ophd.vcxproj]: For reference:
[VCCLCompilerTool.EnablePREfast Property](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.vcprojectengine.vcclcompilertool.enableprefast?view=visualstudiosdk-2017)

> true to enable code analysis (/analyze).

**Comment by cugone** [comment]
The warnings for the Core Guidelines are just too numerous to fix in one go. For now I've just enabled `Microsoft Native Recommended`. It gives good coverage of problem areas and doesn't add much overhead in the way of warning count.

**Comment by cugone** [comment]
Recent warnings fixes and potential additions to the list of warnings to-be-added make this current PR obsolete. Closing to remake.

**Pull request closed without merging**